### ISSUE TITLE
please reference prior art and learn from it

### ISSUE THREAD

**Issue by cgwalters** [rpm-software-management/libpkgmanifest#6]
Presumably, this project was motivated by https://github.com/rpm-software-management/dnf5/issues/833 ? There's a lot of prior art noted in that ticket. It'd be good if this project at least noted that and seemed to have made some effort to analyze those prior things for their positives/negatives.

For instance, I notice this one includes the URL. This ruins "git diff" when one is using datestamped composes: https://github.com/konflux-ci/rpm-lockfile-prototype/pull/18 

Also, in rpm-ostree's lockfile as [used by fedora-coreos-config](https://github.com/coreos/fedora-coreos-config/) we ended up splitting each architecture into a separate file. Notice how much simpler the lockfile there is.

Unrelated to the above:

> checksum: md5:1254011a2ce1245f0b528b1190fa3cb6

md5??? I'm guessing someone just wrote the example as a strawman of the YAML format, but please since you're creating something new don't even try to support md5.

**Comment by praiskup** [comment]
> For instance, I notice this one includes the URL. This ruins "git diff"

I agree that including the full URL for every component would unnecessarily
bloat the lockfiles.  Using a `repo URL` combined with a `relative location`
would likely cover most use cases.  But for Mock which frequently works with
mirrored content.  I filled #7.

**Comment by mcurlej** [comment]
> Also, in rpm-ostree's lockfile as [used by fedora-coreos-config](https://github.com/coreos/fedora-coreos-config/) we ended up splitting each architecture into a separate file. Notice how much simpler the lockfile there is.

To have all the architectures in one file was a requirement. But i am not opposed to have a switch which could do that.

**Comment by jan-kolarik** [comment]
Hi Colin,

> Presumably, this project was motivated by rpm-software-management/dnf5#833 ? There's a lot of prior art noted in that ticket. It'd be good if this project at least noted that and seemed to have made some effort to analyze those prior things for their positives/negatives.

Well, kind of. While addressing the issues in https://github.com/rpm-software-management/dnf5/issues/833 is not the primary objective of this project, we aim to incorporate the key considerations mentioned there. If there is a strong demand, we are open to extending the current solution. It appears there are several stakeholders involved, so we're starting with an MVP that covers basic functionality for all of them.

> For instance, I notice this one includes the URL. This ruins "git diff" when one is using datestamped composes: konflux-ci/rpm-lockfile-prototype#18

Thanks, that's a good point. I’ve tried to address this in the existing https://github.com/rpm-software-management/libpkgmanifest/pull/12.

> Also, in rpm-ostree's lockfile as used by fedora-coreos-config we ended up splitting each architecture into a separate file. Notice how much simpler the lockfile there is.

Currently, the main request was to use a single file. However, adding support for generating separate manifests per architecture could be easily implemented in the dnf plugin, which serves as a client for this library.

> md5??? I'm guessing someone just wrote the example as a strawman of the YAML format, but please since you're creating something new don't even try to support md5.

Yes, this was only intended as an example. I definitely understand that `md5` is deprecated, but since it’s still supported by dnf and rpm, it's just included to ensure compatibility. The support here is more about handling the possibility of md5 values appearing in manifests from older or third-party repositories.

Well, this is really just an example. I know the md5 method is obsoleted now, though it's still supported by dnf and rpm, and the support here simply means that the manifest file can accept the `md5` value in case it shows up unexpectedly from a wild third-party repo. :slightly_smiling_face:

**Comment by jan-kolarik** [comment]
I believe the key points related to project references have been addressed now in the latest commits.

**Comment by cgwalters** [comment]
Reusing this closed ticket as a discussion forum, we can figure out whether to move to a new ticket or a discussion or something else as a followup.

I looked at https://github.com/rpm-software-management/libpkgmanifest/blob/main/docs/design.md and some of the code.

There's a lot of stuff there about code structure and unit tests and such but not much that I can tell that's thinking about the *user experience* of this or precisely (or even semi-precisely) how it would interact with other existing tools.

I mean, I *think* it's trying to replace https://github.com/konflux-ci/rpm-lockfile-prototype but...maybe it's just trying to be a shared library and that project would continue to exist?

It is an *annoying pain* right now in my container builds to have *two* sources of truth for my packages; one in `RUN dnf install` or whatever, and one in the lockfile that I need to keep in sync.

Of course I could write code to parse the lockfile and convert that to something `dnf install` expects, but to me part of the whole point of this is that we're trying to make using lockfiles feel natural, so I'd actually expect dnf to be able to consume this lockfile directly for example.

One exercise you could do that would probably be very helpful would be to try switching over this project to "self hosting" its CI on a container image with locked RPMs. Live that world and step through it - how do you keep that lockfile up to date? Custom Github action? How do you keep what's in the lockfile in sync with all the `BuildRequires` you have to have in the spec file? etc.

**Comment by jlebon** [comment]
> Of course I could write code to parse the lockfile and convert that to something `dnf install` expects, but to me part of the whole point of this is that we're trying to make using lockfiles feel natural, so I'd actually expect dnf to be able to consume this lockfile directly for example.

IIUC, there are dnf plugins in the works for this.

I see https://github.com/rpm-software-management/dnf-plugins-core/commit/03e32be9ebd4402878fbda3c5d37878d213eaaf2 for dnf4 (linked from https://github.com/rpm-software-management/dnf5/issues/833#issuecomment-2358176700).

Is there one for dnf5? We'll need this for FCOS eventually, but I guess for now we can manually re-implement it. Except, argh... I just realized that [`dnf shell` doesn't exist in dnf5](https://github.com/rpm-software-management/dnf5/issues/153), so this puts a cap on your `dnf install` invocation if you want to do this in one transaction. It feels like we at least need a `--argfile`-type argument.

**Comment by cgwalters** [comment]
Ah yes! Good find, I had missed that comment. That is quite helpful indeed. I think we should probably take the high level proposed CLI from that dnf4 plugin to a design doc proposal?
